File: src/io/wende/aoc/seventeen/Three.java
package io.wende.aoc.seventeen;

import io.wende.aoc.common.Puzzle;

public class Three extends Puzzle {

  public Three() {
    super();
    //this.input = "361527";
    this.input = "14";
  }

  public static void main(final String... args) {
    new Three().run();
  }

  private void run() {
    int destination = Integer.parseInt(input);
    int steps = 0;
    for(int i = 1; i * i < destination; i++) {
      out("Step {}", i*i);
      steps = i;
    }



    out("Solution is {}", steps);
  }
}

// 37  36  35  34  33  32  31
// 38  17  16  15  14  13  30
// 39  18   5   4   3  12  29
// 40  19   6   1   2  11  28
// 41  20   7   8   9  10  27
// 42  21  22  23  24  25  26
// 43  44  45  46  47  48  49

File: src/io/wende/aoc/fifteen/Three.kt
package io.wende.aoc.fifteen

import Point
import io.wende.aoc.common.Task

class Three (test: Boolean) : Task(test) {

    companion object {
        var visitedHouses : MutableMap<Point, Int> = mutableMapOf()
        var splitHouses : MutableMap<Point, Int> = mutableMapOf()
    }

    override fun run() {
        val currentPos = Point(0,0)
        visitedHouses[currentPos.snap()] = 1

        this.input[0].map { char -> Direction.from(char.toString()) }
            .forEach { dir ->
                val workingPos = currentPos.walk(dir).snap()
                visitedHouses[workingPos] = visitedHouses.getOrElse(workingPos) { 0 }  + 1
            }

        println("Houses visited: ${visitedHouses.count()}")
        println("Houses visited more than once: ${visitedHouses.count{ entry -> entry.value > 1 }}")

        val currentSanta = Point(0,0)
        val currentRobo = Point(0,0)

        splitHouses[currentSanta.snap()] = 2

        this.input[0].map { char -> Direction.from(char.toString()) }
            .forEachIndexed { i, dir ->
                val workingPos = if(i % 2 == 0) {
                    currentSanta.walk(dir).snap()
                } else {
                    currentRobo.walk(dir).snap()
                }
                splitHouses[workingPos] = splitHouses.getOrElse(workingPos) { 0 } + 1
            }
        println("Houses visited: ${splitHouses.count()}")
        println("Houses visited more than once: ${splitHouses.count{ entry -> entry.value > 1 }}")

    }
}